### ISSUE TITLE
support android arm64 and arm32

### ISSUE THREAD

**Issue by felangel** [shorebirdtech/shorebird#119]
Customers will need appbundles which support both arm64 and arm32 to submit to the stores.  arm32 still accounts for single-digit percentage of Android installs and is thus needed for store submisions.

**Comment by eseidel** [comment]
This will require us to send arch along with artifact creation step.

(edit: Felix has since done this.)

**Comment by eseidel** [comment]
I looked into this a bunch today.  The limiting factor is currently that the way we hook into the `flutter` command is limited to a single platform.  `--local-engine` overrides the defaults in Gradel and makes it single-architecture.  Thus you can't use `flutter build appbundle` to build a Play-store ready appbundle, as you'll only end up with arm64 (or arm32) rather than both.
https://github.com/flutter/flutter/blob/4be45e950dd72ea16e1d0b3387972b9913eefbca/packages/flutter_tools/lib/src/android/gradle.dart#L301

Another way we can hook in is instead to use the `flutter` mirror support https://docs.flutter.dev/community/china FLUTTER_STORAGE_BASE_URL, which can tell Gradel to download our artifacts instead of Flutter's.  I suspect that could end up poisoning the dev's gradle cache though, so we may need to find a way to edit flutter to request different prefixes from our proxy cache rather than the default io.flutter ones:
https://github.com/flutter/flutter/blob/d72d4adbad38c2340ebe2e440a8d0e2a11f88fd9/packages/flutter_tools/gradle/flutter.gradle#L331

**Comment by eseidel** [comment]
There is in fact a gradle build cache (at ~/.gradle/cache on my machine).
https://docs.gradle.org/current/userguide/build_cache.html
And it does seem to contain some the flutter jars.  So if we give a different url and vend modified flutter artifacts with the same name, we do risk poisoning the cache.

https://docs.gradle.org/current/userguide/build_environment.html#sec:gradle_environment_variables

**Comment by eseidel** [comment]
If we wanted to try the proxy approach (again with potential cache poisoning risks) we would need to set up a proxy locally (either use someone elses or write a tiny one) which proxie's requests to localhost/ to download.flutter.io/ and then caches the files and serves those.  That would give us a list of all the files we need and then we can just sub in our .jar files and run it again and see if it correctly builds with our shorebird engine instead.

**Comment by eseidel** [comment]
Another approach could be to fix the Flutter gradle builds to support --local-engine with multiple platforms?
Specifically:
`flutter build appbundle --local-engine-src-path ../engine/src`
or maybe:
`flutter build appbundle --local-engine-src-path ../engine/src --local-engine android_release --local-engine android_release_arm64`
might be better?

I'm not sure if that would be a welcome change or not, but assuming it's possible (we would have to teach it how to assume the correct build directory names?) that would solve the issue w/o cache poisoning risks.

**Comment by eseidel** [comment]
OK, I've come up with a plan.

tl;dr:  fork flutter/flutter (and use that from `shorebird`), just to change the engine.version file to point to our forked flutter/engine hash, which we provide by overriding download.flutter.io via FLUTTER_STORAGE_BASE_URL.

I believe the right thing to do here is for `shorebird` to fork and vend `flutter/flutter`, just like we have forked the flutter/engine and flutter/buildroot already.  That gives us maximal control with minimal impact on Flutter itself.  It's also long-term aligned.  The above solutions are short-term stop gaps which if they hit obstacles result in dead ends, where as if we vend a fork of flutter/flutter we are in complete control.

Once we do that, the change I propose we change the engine.version in `flutter/bin/internal/engine.version` to match our engine version (should also minimize confusion with upstream Flutter if Shorebird-related woes get reported there).  We also then pass FLUTTER_STORAGE_BASE_URL and take responsibility for vending all artifacts.

I don't think we can change the io.flutter prefix to dev.shorebird without potentially breaking other gradle dependencies (e.g. plugins?). So we may just need to rely on the uniqueness of our git hashes from our fork.  If we can change io.flutter to dev.shorebird in the gradle dependencies that could be nice as it may reduce upstream confusion (e.g. when somehow someone takes our fork and runs it normally and it tries to ask for artifacts download.flutter.io doesn't have).

Steps it would involve
- [x] Fork flutter/flutter into shorebirdtech/flutter
- [x] Teach `shorebird` command how to check out `shorebirdtech/flutter` and use that.
- [x] Write a script to capture all needed artifacts (even ones we're not building yet) and upload them to our own storage bucket.  The script needs only accept connections and proxy to download.flutter.io, `flutter precache` should trigger all needed downloads.
- [x] Teach `shorebird` to call our vended Flutter with FLUTTER_STORAGE_BASE_URL pointing to our base url which will of course have our forked/built artifacts (in addition to unmodified copies).  Since we're changing the engine.version we will need to copy artifacts from flutter from base_version to our_version as part of the bucket creation (at least until we replace other artifacts with our versions.  It's possible those other artifacts could also be omitted, since we shouldn't be triggering their need from our forked flutter until we support those platforms anyway?

That should then result in a (forked) `flutter build appbundle` command correctly building an appbundle with both arm32 and arm64 binaries in it.

Is that overkill?  I welcome other proposals.

**Comment by eseidel** [comment]
We should now fully support arm32 Android from an arm64 mac.   The only thing that is still tying us to arm64 is the patch tool, which we either need to vend per-host-platform artifacts for, or need to re-write in Dart.
